conda/cli/main_rename.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+import tempfile
+from contextlib import contextmanager
+from functools import partial
+import os
+
+from ..base.context import context, locate_prefix_by_name, validate_prefix_name
+from ..base.constants import DRY_RUN_PREFIX
+from ..cli import common, install
+from ..common.path import expand
+from ..exceptions import CondaEnvException
+from ..gateways.disk.delete import rm_rf
+from ..gateways.disk.update import rename
+
+
+def validate_src(args) -> str:
+    """
+    Validate that the "base" environment is not being renamed and that the
+    currently active prefix isn't trying to be renamed.
+    """
+    if context.target_prefix == context.root_prefix:
+        raise CondaEnvException("The 'base' environment cannot be renamed")
+
+    prefix = args.name if args.name else args.prefix
+
+    if common.is_active_prefix(prefix):
+        raise CondaEnvException("Cannot rename the active environment")
+
+    return locate_prefix_by_name(prefix)
+
+
+def validate_destination(dest: str, force: bool = False) -> str:
+    """Ensure that our destination does not exist"""
+    if os.sep in dest:
+        dest = expand(dest)
+    else:
+        dest = validate_prefix_name(dest, ctx=context, allow_base=False)
+
+    if not force and os.path.exists(dest):
+        raise CondaEnvException("Environment destination already exists. Override with --force.")
+
+    return dest
+
+
+@contextmanager
+def safe_rm_destination(directory: str):
+    """
+    Used for removing a directory when there are dependent actions (i.e. you need to ensure
+    other actions succeed before removing it).
+
+    Example:
+        with safe_rm_destination(directory):
+            # Do dependent actions here
+    """
+    tmpdir = tempfile.mkdtemp()
+
+    try:
+        rename(directory, tmpdir)
+        yield
+    except Exception as exc:
+        # Error occurred, roll back change
+        rename(tmpdir, directory)
+        raise exc
+
```

Comment:
Can we either make this private (`_safe_rm_destination`), inline it within `execute` below, or abstract it and define it in `conda.gateways.disk.update` (something like `renamecontext`).

I _think_ I prefer the `conda.gateways.disk.update.renamecontext` option since this is very generic functionality and is likely useful elsewhere.

Reply:
Yeah, that thought definitely crossed my mind too. I will refactor it so that it's available to other modules.